### ISSUE TITLE
How to chain failure::Error ?

### ISSUE THREAD

**Issue by rainboh** [withoutboats/failure#176]
I have multiple modules with functions returning failure::Error and I want to chain the returned Errors. It is very simple to assign a std::error::Error to the `#[cause]`, but it's not possible with failure::Error.

To workaround this I used the available failure::Compat and compat(), but then I run into the issue how to print the complete error chain.

This code demonstrates my problem

~~~
#[macro_use]
extern crate failure;
use failure::{Compat, Error, Fail};

mod a {
    use std;
    use failure::{Compat, Error};
    use b;

    #[derive(Fail, Debug)]
    pub enum MyError {
        // #[fail(display = "module a failed")] Failed(#[cause] Error,
        #[fail(display = "module a failed")] Failed(#[cause] Compat<Error>),
    }

    pub fn fails() -> std::result::Result<(), Error> {
        b::fails().map_err(|err| MyError::Failed(err.compat()))?;
        Ok(())
    }
}

mod b {
    use std;
    use failure::Error;
    use std::fs::File;

    #[derive(Fail, Debug)]
    pub enum MyError {
        #[fail(display = "module b failed")] Failed(#[cause] std::io::Error),
    }

    pub fn fails() -> std::result::Result<(), Error> {
        File::open("foo.txt").map_err(|err| MyError::Failed(err))?;
        Ok(())
    }
}

pub fn print_compat_err(err: Error) {
    let mut fail: &Fail = err.cause();
    println!("{}", fail);

    while let Some(cause) = fail.cause() {
        println!("caused by: {}", cause);
        if let Some(compat) = cause.downcast_ref::<Compat<Error>>() {
            fail = compat.inner_ref().cause();
        } else {
            fail = cause;
        }
    }
}

fn print_err(err: Error) {
    let mut fail: &Fail = err.cause();
    println!("{}", fail);

    while let Some(cause) = fail.cause() {
        println!("caused by: {}", cause);
        fail = cause;
    }
}

fn main() {
    if let Err(err) = b::fails() {
        print_err(err);
    }

    if let Err(err) = a::fails() {
        print_err(err);
    }

    if let Err(err) = a::fails() {
        print_compat_err(err);
    }
}
~~~

When printing the error chain with "print_err(err);" then printout for Error returned by "a::fails()" is truncated.

~~~
module a failed
caused by: module b failed
~~~

Printing the error chain with "print_compat_err(err);" displays the expected result.

~~~
module a failed
caused by: module b failed
caused by: No such file or directory (os error 2)
~~~

compat.rs only offers wrapper "into_inner(self)", but for this I need to 'get a reference to the inner error of Compat

~~~
   /// get ref to inner error.
   pub fn inner_ref(&self) -> &E {
       &self.error
   }
~~~

Can this fn be added to the crate?

Or is my use case to return failure::Error in each module not a proper approach.

Any advice and help are welcome.

**Comment by withoutboats** [comment]
This is an interesting problem.

The correct code to *generate* would be to call `.cause()` on the error (or, in #168, `as_fail`). I don't see an easy way to automatically do that, we might need a second attribute for when you want to store the cause as a `failure::Error`.

**Comment by dekellum** [comment]
I was also trying to annotate `#[cause]` for a failure::Error in a variant of a custom `#[derive(Fail)]` enum, and ran into the same original problem described by @rainboh, with compile error given below.  Instead of using `Compat<failure::Error>`, I went ahead and implemented `Display` and `Fail` manually, using `failure::Error::cause()` for my impl. of `Fail::cause` with this variant.   @withoutboats: Is that not the best workaround for the original problem, lacking a means to derive and generate the same?  #168 also suggests "someday when Error can implement Fail" which would presumably also solve this problem?

~~~
error[E0277]: the trait bound `failure::Error: std::error::Error` is not satisfied
  --> src/barc.rs:65:10
   |
65 | #[derive(Fail, Debug)]
   |          ^^^^ the trait `std::error::Error` is not implemented for `failure::Error`
   |
   = note: required because of the requirements on the impl of `failure::Fail` for `failure::Error`
   = note: required for the cast to the object type `failure::Fail`
~~~

**Comment by dekellum** [comment]
@mitsuhiko I was just trying, per above, to get clarification on a best workaround recommendation.  Reasonable request? I don't expect an actual fix was included in 0.1.2, but workaround guidance would be valuable here.

**Comment by mitsuhiko** [comment]
@dekellum i want to look at this after 0.1.2 is out. I'm generally really unhappy with `Error` but i also don't have a good idea yet how to fix it.

**Comment by dekellum** [comment]
Again, just to be clear, I was (original with @withoutboats) just requesting workaround guidance for this issue with `derive(Fail)`. But maybe your response, regarding `Error`, is more directly related than I am currently understanding? Thanks again.

**Comment by mitsuhiko** [comment]
I assume this issue here is effectively a duplicate of #187?

**Comment by dekellum** [comment]
I think that's fair assumption, but just to be clear, I'm looking for a recommend workaround for this variant of the core issue. :+1:

**Comment by mitsuhiko** [comment]
As mentioned there I do not know yet how to deal with this and I do not have a good plan yet. Also obviously no workaround.

**Comment by dekellum** [comment]
I proposed a alternative workaround above (manual impl Fail).  Given the radio silence for 3 months, I'm concerned there might be something wrong with it, and so I'm looking for some comment as to its applicability.

**Comment by dekellum** [comment]
Also, @mitsuhiko, you say "obviously no workaround". This issue proposes two possible workaround: one using `Compat`, one with manual `impl Fail`. They both might be situationally valid? Would it not help users of the new 0.1.2 or earlier releases to see a FAQ entry (possibly just this issue remaining open) with some validation of either or both of those workarounds, or Pros/Cons of each?  Regarding potential pros of the latter workound: its more forward-proof for an when `impl Fail for Error` becomes possible, in that there is no redundant `Compat` wrapper. Thoughts, or am I still missing the larger issue?

**Comment by mitsuhiko** [comment]
> Would it not help users of the new 0.1.2 or earlier releases to see docs or a FAQ entry (possibly just this issue remaining open) with some validation of either or both of those workarounds

I don't think this is the most pressing issue with failure right now and I do not actually want to encourage workarounds for now. IMO users should not be encouraged *for now* to embed errors in errors. I effectively want to postpone dealing with this issue until I managed to wrap my head around it.

**Comment by dekellum** [comment]
FWIW, I think that train has left the station and pretending otherwise is less than helpful, but leaving this issue open would be a reasonable compromise from my perspective. :=)  

"Not encourag[-ing]" is effectively not encouraging use of the *failure* crate, without replacement. But thanks for reading, I feel you've heard my out, and your new work on *failure* is greatly appreciated!
